intermediate_source/FSDP_adavnced_tutorial.rst

```diff
@@ -0,0 +1,534 @@
+            transformer_auto_wrap_policy,
+            transformer_layer_cls={
+                T5Block,
+            },
+        )
+
+        sharding_strategy: ShardingStrategy = ShardingStrategy.FULL_SHARD
+        torch.cuda.set_device(local_rank)
+    
+    
+        init_start_event = torch.cuda.Event(enable_timing=True)
+        init_end_event = torch.cuda.Event(enable_timing=True)
+
+        init_start_event.record()
+
+    
+        model = FSDP(model,
+            auto_wrap_policy=t5_auto_wrap_policy,
+            mixed_precision=bfSixteen,
+            sharding_strategy=sharding_strategy,
+            device_id=torch.cuda.current_device())
+
+        print(model)
+        optimizer = optim.Adadelta(model.parameters(), lr=args.lr)
+
+        scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
+        best_val_loss = float("inf")
+        curr_val_loss = float("inf")
+        file_save_name = "3B-model-"
+
+        if rank == 0:
+            time_of_run = get_date_of_run()
+            dur = []
+            train_acc_tracking = []
+            val_acc_tracking = []
+            training_start_time = time.time()
+
+        if rank == 0 and args.track_memory:
+            fn = "memory_tracking.txt"
+            mem_alloc_tracker = []
+            mem_reserved_tracker = []
+
+        for epoch in range(1, args.epochs + 1):
+            t0 = time.time()
+            train_accuracy = train(args, model, rank, world_size, train_loader, optimizer, epoch, sampler=sampler1)
+            if args.run_validation:
+                curr_val_loss = validation(model, rank, world_size, val_loader)
+            scheduler.step()
+            
+            if rank == 0:
+
+                print(f"--> epoch {epoch} completed...entering save and stats zone")
+
+                dur.append(time.time() - t0)
+                train_acc_tracking.append(train_accuracy.item())
+
+                if args.run_validation:
+                    val_acc_tracking.append(curr_val_loss.item())
+
+                if args.track_memory:
+                    mem_alloc_tracker.append(
+                        format_metrics_to_gb(torch.cuda.memory_allocated())
+                    )
+                    mem_reserved_tracker.append(
+                        format_metrics_to_gb(torch.cuda.memory_reserved())
+                    )
+
+        init_end_event.record()
+
+        if rank == 0:
+            print(f"Cuda event elapsed time: {init_start_event.elapsed_time(init_end_event) / 1000}sec")
+            print(f"{model}")
+
+        if args.save_model and curr_val_loss < best_val_loss:
+
+            # save
+            if rank == 0:
+                print(f"--> entering save model state...")
+            save_policy = FullStateDictConfig(offload_to_cpu=True, rank0_only=True)
+            with FSDP.state_dict_type(
+                model, StateDictType.FULL_STATE_DICT, save_policy
+            ):
+                cpu_state = model.state_dict()
```

Comment:
possibly add optimizer state checkpointing as well?

Reply:
@rohan-varma it is waiting for the final fix yet.